### PR TITLE
fix(DatePicker): ability to pass custom parser for input value

### PR DESCRIPTION
**Pull request by sanex3339**
[SPT-2161]

### Description

Describe the changes and motivations for the pull request.

### How to test

- FIXME: Add the steps describing how to verify your changes

### Screenshots

| Before.                                 | After.                                  |
| --------------------------------------- | --------------------------------------- |
| Insert screenshots or screen recordings | Insert screenshots or screen recordings |

### Review

- [ ] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/api-principles.md)
- [ ] Annotate all `props` in component with documentation
- [ ] Create `examples` for component
- [ ] Ensure that deployed demo has expected results and good examples
- [ ] Ensure that tests pass by running `yarn test`
- [ ] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [ ] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and
1. You have no extra requests.
2. You have optional requests.
   1. Add `nit: ` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because
1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:
1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[SPT-2161]: https://toptal-core.atlassian.net/browse/SPT-2161?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sanex3339** [comment]
@toptal-bot run package:alpha-release

**Comment by sanex3339** [comment]
@toptal-bot run package:alpha-release

**Comment by sanex3339** [comment]
@toptal-bot run package:alpha-release

**Comment by sanex3339** [comment]
@toptal-bot run package:alpha-release

**Comment by TomasSlama** [comment]
@toptal-bot run package:alpha-release

**Comment by TomasSlama** [comment]
@toptal-bot run visual

File: packages/picasso-lab/src/DatePicker/DatePicker.tsx [hunk 1]
```diff
@@ -89,6 +92,8 @@ export interface Props
   error?: boolean
   /** Function to override default markup to show Date */
   renderDay?: (args: DayProps) => ReactNode
+  /** Indicate if human-readable dates should be parsed inside `DatePicker`'s input */
+  parseHumanReadableDates?: boolean
```
Comment (dmaklygin1) [inline packages/picasso-lab/src/DatePicker/DatePicker.tsx:96]: @MrBra1nwash added `parseInputValue` prop to parse input value. 
If I'm not mistaken, you're trying to achieve the same goal: parse input value here.

File: packages/picasso-lab/src/DatePicker/DatePicker.tsx [hunk 2]
```diff
@@ -208,6 +215,33 @@ export const DatePicker = (props: Props) => {
...
+    const nextInputValue = event.currentTarget.value
+
+    const isValidDate = isDateValid(nextInputValue, editDateFormat)
+    const shouldParseHumanReadableDate =
+      parseHumanReadableDates && !range && nextInputValue && !isValidDate
+
+    if (shouldParseHumanReadableDate) {
+      setIsInputLoading(true)
+
+      const parsedHumanReadableDate = await datePickerParseHumanReadableDateString(
```
Comment (dmaklygin1) [inline packages/picasso-lab/src/DatePicker/DatePicker.tsx:230]: Not against this solution, but what if we want to expose Roman numerals, should we also need to add a function to this library?

Comment (dmaklygin1) [reply packages/picasso-lab/src/DatePicker/DatePicker.tsx:230]: I would pursue more extensive approach in DatePicker

File: packages/picasso-lab/src/DatePicker/DatePicker.tsx [hunk 3]
```diff
@@ -311,6 +345,16 @@ export const DatePicker = (props: Props) => {
       </InputAdornment>
     ) : undefined
 
+  const loadingComponent = (
+    <InputAdornment
+      data-testid='loading-adornment'
```
Comment (denieler) [inline packages/picasso-lab/src/DatePicker/DatePicker.tsx:350]: hardcoded data-testId 🔴

Comment (sanex3339) [reply packages/picasso-lab/src/DatePicker/DatePicker.tsx:350]: It's POC. For the final solution we will fix it

File: packages/picasso-lab/src/DatePicker/DatePicker.tsx (hunk 1 above)
Comment (sanex3339) [reply packages/picasso-lab/src/DatePicker/DatePicker.tsx:96]: `parseInputValue` seems just a temporary solution just to add support of strings (only as a Date type) with `human readable dates` to the DatePicker.

After pure FE implementation on the `picasso` side, most likely `parseInputValue` can be removed.

File: packages/picasso-lab/src/DatePicker/DatePicker.tsx (hunk 2 above)
Comment (sanex3339) [reply packages/picasso-lab/src/DatePicker/DatePicker.tsx:230]: Discussed.  I will rework this POC with ability to pass `custom` parsers

File: packages/picasso-lab/src/DatePicker/DatePicker.tsx [hunk 4]
```diff
@@ -219,22 +232,33 @@ export const DatePicker = (props: Props) => {
...
+      parseInputValue && !isValidDate
+        ? datePickerParseCustomDateString(
+            nextValue,
+            parseInputValue,
+            parserOptions
+          )
+        : datePickerParseDateString(nextValue, parserOptions)
+
+    if (parsedInputDate) {
+      onChange(parsedInputDate)
```
Comment (dmaklygin1) [inline packages/picasso-lab/src/DatePicker/DatePicker.tsx:261]: as considered, the whole approach is fine, but let's think about this construction inside `handleInputChange` a little bit more: how can we improve it

**Review by dmaklygin1 — COMMENTED** [review]
Good approach, I like it!

File: packages/picasso-lab/src/DatePicker/utils.ts [hunk 5]
```diff
@@ -118,6 +126,39 @@ export const datePickerParseDateString: DatePickerStringParser = (
...
+
+  const nextTimezoneValue = timezoneFormat(parsedDate, timezone)
+
+  if (!isDateWithinInterval(nextTimezoneValue, minDate, maxDate)) {
+    return
+  }
+
+  return parsedDate
+}
+
```
Comment (dmaklygin1) [inline packages/picasso-lab/src/DatePicker/utils.ts:161]: let's think about `datePickerParseDateString ` and `datePickerParseCustomDateString`. I hope, we may improve them

File: packages/picasso-lab/src/DatePicker/DatePicker.tsx (hunk 2 above)
Comment (sanex3339) [reply packages/picasso-lab/src/DatePicker/DatePicker.tsx]: Updated

File: packages/picasso-lab/src/DatePicker/utils.ts (hunk 5 above)
Comment (sanex3339) [reply packages/picasso-lab/src/DatePicker/utils.ts]: https://github.com/toptal/picasso/pull/2277/commits/ccecf6b5b64b98212bf261f29c4e01c0724dda89

File: packages/picasso-lab/src/DatePicker/DatePicker.tsx (hunk 4 above)
Comment (sanex3339) [reply packages/picasso-lab/src/DatePicker/DatePicker.tsx:250]: https://github.com/toptal/picasso/pull/2277/commits/ccecf6b5b64b98212bf261f29c4e01c0724dda89

**Comment by sanex3339** [comment]
@toptal-bot run package:alpha-release

**Comment by sanex3339** [comment]
POC with integration to staff portal:
https://github.com/toptal/staff-portal/pull/4892

**Pull request closed without merging**